### ISSUE TITLE
Overlay not closed when clicking another MenuTrigger

### ISSUE THREAD

**Issue by madmanwithabike** [adobe/react-spectrum#1155]
This is possibly related to #850, but using the @react-aria hooks only and is repeatable on every click, #850 seems to be fast click only and not reproducible at all in some cases.

# 🐛 Bug Report 

Clicking between two (or more) buttons with `useMenuTrigger` and `useOverlay`,  does not close the currently open menu.

## 🤔 Expected Behavior

Open menu should close and new menu should open.

## 😯 Current Behavior

This seems to only happen with `useMenuTrigger`, clicking from an open menu to a button using `useButton` closes the menu as expected.

Screenshot:

![image](https://user-images.githubusercontent.com/32389848/95609622-e166ad80-0a56-11eb-88d2-b24b111bf928.png)

When clicking away repeatedly, the overlays are then closed one at a time in reverse order.

## 💁 Possible Solution

Possibly related to #850.

## 💻 Code Sample

Can be reproduced using the useMenuTrigger sample code: https://react-spectrum.adobe.com/react-aria/useMenuTrigger.html#example but duplicating the MenuButton:

```
<MenuButton label="Actions">
  <Item key="copy">Copy</Item>
  <Item key="copy">Copy</Item>
  <Item key="copy">Copy</Item>
</MenuButton>
<MenuButton label="Actions">
  <Item key="copy">Copy</Item>
  <Item key="copy">Copy</Item>
  <Item key="copy">Copy</Item>
</MenuButton>
```

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| @react-aria/button | 3.2.2 |
| @react-aria/menu | 3.1.2 |
| @react-aria/overlays | 3.4.0  |
| Browser          | Chrome + Firefox |
| Operating System | Ubuntu 18.04 + Windows 10 |

Thanks!

**Comment by LFDanLu** [comment]
Reproduced this locally, initial investigation points to this being an issue with how `useOverlay` adds new overlays to its `visibleOverlays` before calling `onInteractOutside`, resulting in `onClose` never being called for the first overlay when the second menu button is clicked.

**Comment by snowystinger** [comment]
Initial thoughts and recommendations

Problem statement: How to know which overlay should be removed, they can be siblings or nested.

Goals:
If siblings, the already open one should be closed and then the new one opened. Order of operations may not matter.
If nested, then already open should stay open and the new one should also open.

Data Structures options:
- `[[modal1, nestedinmodal1], [sibling]]` represents as nested structure, only one internal array, but there may be a transition period in which there are two
- `[[modal1, sibling], [nestedinmodal1]]` represents as layers, only one element should be in a layer at a time, but there may be a transition period in which there are two
- 
  ```
  {
  root: [modal1, sibling],
  modal1: [nestedinmodal1],
  sibling: [],
  nestedinmodal1: []
  }
  ```
  or to take the layers a little further, it could be a doubly linked graph, usually it'd look like a doubly linked list, but there may be a brief period where it's a tree with branches

How to tell if we're in the nested or sibling case
* through context
  * Overlay.tsx? then non rsp users will need to include a new element or use a new hook, they might forget
* all overlays always close, unless a param is passed in that puts it into a controlled state to keep it open
  * probably a breaking change, also, very gross
* look up dom tree
  * nothing to really search for exists right now given that they might or might not portal out and may not aria-hide the rest of the tree can't look at the e.target to see if it's in an overlay, the container(s) may not say that they are an overlay such as in useMenuTrigger example
  * we might be able to pass a prop (data attribute?) back from useOverlay that is  part of the spread though so that we can search for that, this eliminates the need for people to do anything to gain the new functionality, just update useOverlay


To determine if we're in a nested modal or in a sibling on the same layer, I would recommend the data-attribute that we can search the dom for, it's the least impact on users and non-breaking because the change occurs in one package, so as soon as it is updated, they get the change everywhere.
I think the dl tree is probably the best option for the data structure used, for example:
2 menus one with a nested menu which is open all the way. User clicks on the second menu, we can close the first sibling by going down the linked path and then close as we come back up. It's an easy way to see what the order of operations should be.

Other issues:
This change will need to be done after the remove Pointer Events change. Otherwise with mixed pointer/mouse events, there is a bug that onInteractOutside won't be called if the user clicks on anything using usePress.

**Comment by snowystinger** [comment]
Upon initial conversation with team, evaluated that all of the above are complex solutions. We'd like something simpler. To do so, we're going back to the original problem, two menu's open at the same time.
We tried fixing it with https://github.com/adobe/react-spectrum/pull/925
This was reverted by https://github.com/adobe/react-spectrum/pull/980
This failed because two menu's open at the same time was still possible. This is because ...researching again, will fill in as I learn/remember more...
